### ISSUE TITLE
Cookie mage-messages gets polluted on post error and will exceed max cookie size

### ISSUE THREAD

**Issue by Morgy93** [magento/magento2#33611]
### Preconditions (*)

- Magento 2.4.2
- PHP 7.4

### Steps to reproduce (*)

1. Send any post request to the shop (Ex: POST    http://yourmagentostore.com) . Status code: 200
2. See that Cookies - mage-messages is growing on every request with the same error message.
3. On hitting the same request multiple times, getting 500 Internal server error: (ex: say around after 30 requests)
main.CRITICAL: Unable to send the cookie. Size of 'mage-messages' is 4155 bytes.
### Expected result (*)

That the same error is not appended on every request.

### Actual result (*)

```
mage-messages=%5B%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%5D;
```
On hitting the same request multiple times, getting 500 Internal server error:
![image](https://user-images.githubusercontent.com/60198231/127649835-d8f149ad-cbfa-434f-8555-40f102752985.png)
Go to Magento root/var/log/exception.log -
**main.CRITICAL: Unable to send the cookie. Size of 'mage-messages' is 4155 bytes.**
![image](https://user-images.githubusercontent.com/60198231/127649905-d497686f-eba2-42f1-9560-6ab03a7cd2af.png)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by Morgy93** [comment]
@magento give me 2.4-develop instance

**Comment by Morgy93** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible:
Steps to reproduce:
1. `POST shop.tld` ( e.g. `POST https://8f931add3f9319688577122399086a45-2-4-develop.instances.magento-community.engineering/`)
2. See `mage-messages` growing with same error message on each request:
```
Set-Cookie: PHPSESSID=xyzyoloeb37q4jnhvdd; expires=Wed, 28-Jul-2021 09:27:08 GMT; Max-Age=3600; path=/; domain=8f931add3f9319688577122399086a45-2-4-develop.instances.magento-community.engineering; HttpOnly; SameSite=Lax,private_content_version=9416418948c9bf681903aec2394af71a; expires=Sat, 26-Jul-2031 08:27:08 GMT; Max-Age=315360000; path=/; secure; SameSite=Lax,mage-messages=%5B%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22
Invalid+Form+Key.+Please+refresh+the+page.%22%7D%5D; expires=Thu, 28-Jul-2022 08:27:08 GMT; Max-Age=31536000; path=/; SameSite=Strict
```

Example POST:
```
POST https://8f931add3f9319688577122399086a45-2-4-develop.instances.magento-community.engineering/
Content-Type: application/json

{
    "yolo": "ok"
}
```

If you do this often enough (like bots/"hackers" for example) you'll be left with error 500: `Unable to send the cookie. Size of 'mage-messages' is %size bytes.` if it exceedes the configured `%size`. (4107 Bytes seems to be default)

**Comment by engcom-November** [comment]
Verified the issue on Magento 2.4-develop branch and the issue is reproducible:
**Steps to reproduce:**
1. Send any post request to the Magento store:
Ex: POST: http://yourmagentostore.com/
Ex: Request payload:
```
{
    "yolo": "ok"
}
```
Status: 200 OK
**Issue: Cookies - mage-messages  - same message is getting appended on every request.**
mage-messages=%5B%7B%22type%22%3A%22error%22%2C%22text%22%3A%22Invalid%20Form%20Key.%20Please%20refresh%20the%20page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22Invalid%20Form%20Key.%20Please%20refresh%20the%20page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22Invalid%20Form%20Key.%20Please%20refresh%20the%20page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22Invalid%20Form%20Key.%20Please%20refresh%20the%20page.%22%7D%2C%7B%22type%22%3A%22error%22%2C%22text%22%3A%22Invalid%20Form%20Key.%20Please%20refresh%20the%20page.%22%7D%5D; Path=/; Expires=Sat, 30 Jul 2022 10:04:11 GMT;

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-November 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42946` were created

**Issue Available**: @engcom-November, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by giacmir** [comment]
Hit this today in 2.4.3-p2

**Comment by rutgerrademaker** [comment]
Also got hit by this on 2.4.3-p3, figured out there is a MQP to fix this though: MDVA-41215

https://support.magento.com/hc/en-us/articles/4417378995853-MDVA-41215-Users-get-500-error-after-setting-mage-messages-cookie

After applying that the issue could not be reproduced.

**Comment by hostep** [comment]
Thanks @rutgerrademaker

It looks like the same fixes from that patch got included in Magento 2.4.4 by: [ACP2E-290: [Magento Cloud] NewRelic Alert: 'Error Rate Critical'](https://github.com/magento/magento2/commit/44e055e27a50b70093923542d923cec64b45dc6f)

So this issue can probably be closed now?
